### PR TITLE
fix(cgen): run-time heap corruption with static `seq`

### LINKED ISSUE #1117: incorrect codegen for static Table causing corrupted heap
### Example

```nim
import std/tables

let x = static: {"a": "b"}.toTable
echo x
```

### Actual Output (run with AddressSanitizer)

```
=================================================================
==443926==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60d0000000c8 at pc 0x000000418338 bp 0x7ffdcb5f5900 sp 0x7ffdcb5f58f8
WRITE of size 8 at 0x60d0000000c8 thread T0
    #0 0x418337 in NimMainModule /home/leorize/documents/sources/nimskull/test.nim:3
    #1 0x4187ca in NimMain /home/leorize/documents/sources/nimskull/test.nim:1
    #2 0x40141c in main /home/leorize/documents/sources/nimskull/test.nim:9
    #3 0x7f6448e46149 in __libc_start_call_main (/lib64/libc.so.6+0x28149) (BuildId: 788cdd41a15985bf8e0a48d213a46e07d58822df)
    #4 0x7f6448e4620a in __libc_start_main_impl (/lib64/libc.so.6+0x2820a) (BuildId: 788cdd41a15985bf8e0a48d213a46e07d58822df)
    #5 0x401494 in _start (/home/leorize/documents/sources/nimskull/test+0x401494) (BuildId: 22803c7c3d0f26f0e806d23a58d9570a6e8c54ff)

0x60d0000000c8 is located 0 bytes after 136-byte region [0x60d000000040,0x60d0000000c8)
allocated by thread T0 here:
    #0 0x7f64490d8cc7 in calloc (/lib64/libasan.so.8+0xd8cc7) (BuildId: 7fcb7759bc17ef47f9682414b6d99732d6a6ab0c)
    #1 0x407e1d in alloc0Impl__system_1722 /home/leorize/documents/sources/nimskull/lib/system/mm/malloc.nim:8
    #2 0x407e1d in alignedAlloc0__system_1892 /home/leorize/documents/sources/nimskull/lib/system/memalloc.nim:361
    #3 0x407e1d in newSeqPayload /home/leorize/documents/sources/nimskull/lib/system/seqs_v2.nim:38
    #4 0x417770 in NimMainModule /home/leorize/documents/sources/nimskull/test.nim:3
    #5 0x4187ca in NimMain /home/leorize/documents/sources/nimskull/test.nim:1
    #6 0x40141c in main /home/leorize/documents/sources/nimskull/test.nim:9
    #7 0x7f6448e46149 in __libc_start_call_main (/lib64/libc.so.6+0x28149) (BuildId: 788cdd41a15985bf8e0a48d213a46e07d58822df)
    #8 0x7f6448e4620a in __libc_start_main_impl (/lib64/libc.so.6+0x2820a) (BuildId: 788cdd41a15985bf8e0a48d213a46e07d58822df)
    #9 0x401494 in _start (/home/leorize/documents/sources/nimskull/test+0x401494) (BuildId: 22803c7c3d0f26f0e806d23a58d9570a6e8c54ff)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/leorize/documents/sources/nimskull/test.nim:3 in NimMainModule
Shadow bytes around the buggy address:
  0x60cffffffe00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x60cffffffe80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x60cfffffff00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x60cfffffff80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x60d000000000: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
=>0x60d000000080: 00 00 00 00 00 00 00 00 00[fa]fa fa fa fa fa fa
  0x60d000000100: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x60d000000180: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x60d000000200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x60d000000280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x60d000000300: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==443926==ABORTING
```

### Expected Output

```
{"a": "b"}
```

### Additional Information

* The culprit is wrong codegen for `Table.data` seq:

```c
	T2_.len = 8;
        T2_.p = (tySequence__SFipDZIkFtPcrLs0AGBBjQ_Content*)
         newSeqPayload(
           8,
           sizeof(tySequence__SFipDZIkFtPcrLs0AGBBjQ),
           NIM_ALIGNOF(tySequence__SFipDZIkFtPcrLs0AGBBjQ));
	T2_.p->data[0] = _1;
	T2_.p->data[1] = _2;
	T2_.p->data[2] = _3;
	T2_.p->data[3] = _4;
	T2_.p->data[4] = _5;
	T2_.p->data[5] = _6;
	T2_.p->data[6] = _7;
	T2_.p->data[7] = _8;
	_10 = T2_;
	eqcopy___test_146((&_9), _10);
	nimZeroMem((void*)(&_11), sizeof(tyObject_Table__9bvsDyXFoMFA19bQi5WBXMdA));
	_11.data = _9;
```

(*code has been modified for readability*)

Instead of allocating with the size of `KeyValuePair` tuple, this code allocates with the size of the seq itself, as such the allocated memory region has incorrect size, causing heap corruption.

### PR DESCRIPTION
**Pull request by zerbina**
## Summary

Fix a C code generator bug where using the result of a `static`-
evaluated expression (i.e., `static: x`), resulted in a heap corruption
at run-time. Only values that had a generic `seq` alias as their type
were affected.

Fixes https://github.com/nim-works/nimskull/issues/1117.

## Details

The problem is that `genSeqConstr` didn't call `skipTypes` on the
`cnkArrayConstr` (a `seq` construction, in this case) expression. This
works if the type is a `tySeq`, but for `tyGenericInst`, `t.lastSon`
returns the `tySeq`, and treating the `tySeq` as the element is wrong.

Since the size and alignment of the `seq` type itself are then passed
to `newSeqPayload`, to little space is allocated, causing an out-of-
bounds memory access if the actual elements' size is larger than
`sizeof(seq[T])`.

`genSeqConstr` now properly skips the type, fixing the bug.
`genArrToSeq` was also missing the type skipping, it's adjusted too.

### CODE REVIEW

**Comment by zerbina** [comment]
The test failure is because of a separate issue with destructor injection, where the temporary is erroneously copied (seemingly because it's treated as not having a destructor).

I'll look into this tomorrow.

File: tests/ccgbugs/tseq_construction_bug.nim [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    Reduced from https://github.com/nim-works/nimskull/issues/1117"
+  '''
+  targets: "c"
+  matrix: "-d:useMalloc"
+  valgrind: true
+"""
+
+type Sequence[T] = seq[T]
+
+# tcompile-time-evaluated value must have a generic instantiation as the type
```
Comment (alaviss) [inline tests/ccgbugs/tseq_construction_bug.nim]:
```suggestion
# compile-time-evaluated value must have a generic instantiation as the type
```

typo?

Comment (zerbina) [reply tests/ccgbugs/tseq_construction_bug.nim]: Ah, no, I erroneously cut off parts of the "the" when shortening the text.
```suggestion
# the compile-time-evaluated value must have a generic instantiation as the type
```

**Comment by zerbina** [comment]
The minimal example for reproducing the leak:
```nim
var s = static: @[(1, 2)]
```

The problem is that `seq` constructions resulting from `nkBracket` expressions are treated as constants during the MIR phase, and constants need to be copied. This is not a problem for, e.g., `static: @[1, 2]`, which `cgen` lifts into a C constant that doesn't require destruction (and indeed requires copying), but it is a problem for sequences of aggregate type, where `mirgen` translates the construction in a way such that `cgen` cannot lift the expression into a C constant.

Once #1132 is merged, I'll implement `mirgen`-driven lifting of constant expression, which will, as a side-effect, resolve this problem.

**Comment by zerbina** [comment]
I've merged `devel`, which has the fix for the memory leak. As promised, the `tseq_construction_bug` test now works.

**Pull request merged**